wordpress-mobile/WordPress-Android
# Email encoding fix with RFC 3986

Fixes #15480

Related PRs:
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2171
- https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/73

## Description 

Encodes delimiters in email addresses when calling the `auth-options` endpoint, based on the [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2) specification.

See: [Similar fix in iOS](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/457).

More info in this issue comment: https://github.com/wordpress-mobile/WordPress-Android/issues/15480#issuecomment-1202342661.

**To Test:**

**Setup** (feel free to skip if you already have a passwordless account)

0. For all steps, utilize an email with an apostrophe (such as `test'email@gmail.com`) and make sure you don't get an error when attempts are made to log in or sign up.
1. Log out if you are logged in.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is _not_ associated with a WordPress.com account.
1. Complete the sign-up flow normally.

**Test 1: Passwordless account / Magic Link login**

0. Clear app data.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is associated with a passwordless WordPress.com account.
1. Tap **Continue**.
1. Notice the Login Magic Link Sent screen.
1. Notice how there's no option to login with a password.
1. Tap **Check email**.
1. On the login email sent, tap **Log in to the app**.
1. Complete the login flow normally.

**Test 2: Account with password / Password login**

0. Clear app data.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is associated with a WordPress.com account that has a password set.
1. Tap **Continue**.
1. Notice the Password screen.
1. Notice how there's an option to login with a magic link.
1. Enter the WordPress.com account password.
1. Tap **Continue**.
1. Complete the login flow normally.

**Test 3: Sign up with an email not associated with a WordPress.com account**
0. Clear app data.
1. On the Prologue Screen, tap **Log In**.
2. If the Smart Lock dialog appears, dismiss it.
3. On the Email Screen, enter an email address that **_is not_** associated with a WordPress.com account.
4. Tap **Next**.
5. Notice the Sign-Up Magic Link screen.
6. Check email inbox of address used in Step 3.
7. Notice signup email from WordPress.com.
8. Complete the sign-up flow normally.


**Test 4: Check iOS & Android are in sync for emails having a `'` (optional)**
1. Create an account on an iOS with an e-mail having a single quote (E.g.: `main_email+test`ios@gmail.com`)
2. Login on iOS with that email
   a. Check under `Me` → `Account Settings` that the e-mail address is displayed correctly.
3. Login on Android with the same email
   a. **Expect** a successful login
   b. Check under `Me` → `Account Settings` that the e-mail address is displayed correctly.
4. Repeat steps 1-3 creating an account on `Android` first (E.g.: `main_email+test`android@gmail.com`)
   a. **Expect** same results.

## Previews

<details>
<summary>Before</summary>

https://user-images.githubusercontent.com/1509205/140247809-fa135c81-4557-427e-9308-2ec71f4198eb.mov
</details>

<details open>
<summary>After</summary>

https://user-images.githubusercontent.com/4588074/182418659-8275dae2-6dd8-4bb1-8352-e27c29debe18.mp4
</details>

## Merging Strategy

**Notice** ✅ One review should be enough to merge these PRs.

<details>
<summary>📂 Click to reveal merging workflow</summary>

1. Merge the  `WordPress-FluxC-Android` PR first
2. Wait for the [BuildKite Job for trunk](https://buildkite.com/automattic/wordpress-fluxc-android) to complete
   - Then go to the Job details in BuildKite and open subjob with name `Publish :fluxc`
   - Search for `is succesfully published` and **Copy** the value before that text, looking like  `trunk-{commit_sha}`.
   - **Paste** this value in the `./WordPressLoginFlow/build.gradle` file of `WordPress-Login-Flow-Android`, at:
```diff
@@ -55,7 +55,7 @@ dependencies {

api 'com.google.android.gms:play-services-auth:18.1.0'

-    implementation("org.wordpress:fluxc:2171-bac4c0dab01c00b98c2e88b7421ee5c84f408439") {
+    implementation("org.wordpress:fluxc:trunk-{commit_sha}") {
exclude group: "com.android.support"
exclude group: "org.wordpress", module: "utils"
}
```

3. Commit & push the change to the `WordPress-Login-Flow-Android` PR
4. Merge the `WordPress-Login-Flow-Android` PR
5. Also **Paste** the same FluxC version in the `./build.gradle` file of `WordPress-Android`, for `ext.fluxCVersion`.
6. Wait for the [BuildKite Job for trunk](https://buildkite.com/automattic/wordpress-login-flow-android) to complete
   - Then go to the Job details in BuildKite and open subjob with name `Publish to s3`
   - Search for `is succesfully published` and **Copy** the value before that text, looking like  `trunk-{commit_sha}`.
   - **Paste** this value in the `./build.gradle` file of `WordPress-Android`, for `ext.wordPressLoginVersion`.

7. Commit & push this change to the `WordPress-Android` PR
8. Merge the `WordPress-Android` PR.
</details>


## Regression Notes

1. Potential unintended areas of impact
Sign up and login-related API endpoints that provide authentication-based services.

6. What I did to test those areas of impact (or what existing automated tests I relied on)
I ran the login test cases that are shared above. The [SignUp](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/androidTest/java/org/wordpress/android/e2e/SignUpTests.java) and [LogIn](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/androidTest/java/org/wordpress/android/e2e/LoginTests.java) e2e tests were leveraged.

7. What automated tests I added (or what prevented me from doing so)
I didn't add any automated tests to cover this case since the fix is an optimization for the single quote edge case

PR submission checklist:
